File: src/test/java/br/com/blz/testjava/application/product/update/UpdateProductUseCaseTest.java
package br.com.blz.testjava.application.product.update;

import br.com.blz.testjava.application.UseCaseTest;
import br.com.blz.testjava.application.product.InventoryInput;
import br.com.blz.testjava.domain.product.Product;
import br.com.blz.testjava.domain.product.ProductGateway;
import br.com.blz.testjava.domain.product.Sku;
import br.com.blz.testjava.domain.product.Warehouse;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import java.util.List;
import java.util.Optional;

import static java.util.Arrays.asList;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;

public class UpdateProductUseCaseTest extends UseCaseTest {

    @InjectMocks
    private DefaultUpdateProductUseCase useCase;

    @Mock
    private ProductGateway productGateway;

    @Override
    protected List<Object> getMocks() {
        return List.of(productGateway);
    }

    @Test
    public void givenAValidCommand_whenCallsUpdateProduct_shouldUpdatedProduct() {
        final var sku = 43264L;
        final var expectedQuantity = 10L;
        final var expectedIsMarketable = true;
        final var expectedName = "L'Oréal Professionnel Expert Absolut Repair Cortex Lipidium - Máscara de Reconstrução 500g";
        final var warehouses = asList(Warehouse.newWarehouse("SP", 8, "ECOMMERCE"),
            Warehouse.newWarehouse("MOEMA", 2, "PHYSICAL_STORE"));
        final var productCommand = UpdateProductCommand.with(sku, expectedName, InventoryInput.with(warehouses));
        final var expectedSku = Sku.from(sku);
        final var product = Product.newProduct(sku, expectedName, warehouses);

        when(productGateway.findBySku(eq(expectedSku))).thenReturn(Optional.of(product.clone()));
        when(productGateway.update(any())).thenReturn(product);

        final var actualProduct = useCase.execute(productCommand);

        Assertions.assertEquals(expectedSku.getValue(), actualProduct.sku());
        Assertions.assertEquals(expectedName, actualProduct.name());
        Assertions.assertEquals(expectedQuantity, actualProduct.inventory().quantity());
        Assertions.assertEquals(expectedIsMarketable, actualProduct.isMarketable());
    }
}

File: src/test/java/br/com/blz/testjava/domain/usecase/UpdateProductUseCaseTest.kt
package br.com.blz.testjava.domain.usecase

import br.com.blz.testjava.domain.entity.UpdateProduct
import br.com.blz.testjava.domain.exception.NotFoundException
import br.com.blz.testjava.domain.repository.ProductRepository
import br.com.blz.testjava.domain.repository.WarehouseRepository
import br.com.blz.testjava.factory.MockFactory
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.doReturn
import org.mockito.junit.jupiter.MockitoExtension
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

@ExtendWith(MockitoExtension::class)
class UpdateProductUseCaseTest {
  @Mock private lateinit var productRepository: ProductRepository
  @Mock private lateinit var warehouseRepository: WarehouseRepository

  @InjectMocks private lateinit var updateProductUseCase: UpdateProductUseCase

  @Test
  fun `Should update product with correct values for isMarketable and quantity`() {
    val product = MockFactory.createProduct()
    val warehouse = MockFactory.createWarehouse()
    doReturn(product).`when`(productRepository).findById(product.sku.toString())
    doReturn(product).`when`(productRepository).save(product)
    doReturn(warehouse).`when`(warehouseRepository).save(warehouse, product.sku)

    val response = updateProductUseCase.execute(
      UpdateProduct(product.sku, product.name, listOf(
        warehouse,
        warehouse,
      ))
    )

    assertTrue { response.isMarketable }
    assertEquals(10, response.inventory.quantity)
    Mockito.verify(warehouseRepository, Mockito.times(1)).deleteBySku(product.sku)
  }

  @Test
  fun `Should update product with correct values for isMarketable and quantity when is not marketable`() {
    val product = MockFactory.createProduct()
    val warehouse = MockFactory.createWarehouse(quantity = 0)
    doReturn(product).`when`(productRepository).findById(product.sku.toString())
    doReturn(product).`when`(productRepository).save(product)
    doReturn(warehouse).`when`(warehouseRepository).save(warehouse, product.sku)

    val response = updateProductUseCase.execute(
      UpdateProduct(product.sku, product.name, listOf(
        warehouse,
        warehouse,
      ))
    )

    assertFalse { response.isMarketable }
    assertEquals(0, response.inventory.quantity)
  }

  @Test
  fun `Should throw not found exception when sku is not found`() {
    val product = MockFactory.createProduct()
    val warehouse = MockFactory.createWarehouse(quantity = 0)

    assertThrows<NotFoundException> {
      updateProductUseCase.execute(
        UpdateProduct(product.sku, product.name, listOf(
          warehouse,
          warehouse,
        ))
      )
    }
    Mockito.verify(warehouseRepository, Mockito.times(0)).deleteBySku(product.sku)
  }

}
